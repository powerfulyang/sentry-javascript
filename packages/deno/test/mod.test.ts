import { assertEquals } from 'https://deno.land/std@0.202.0/assert/assert_equals.ts';
import { assertSnapshot } from 'https://deno.land/std@0.202.0/testing/snapshot.ts';

import type { sentryTypes } from '../build-test/index.js';
import { sentryUtils } from '../build-test/index.js';
import { DenoClient, Hub, Scope, defaultIntegrations } from '../build/index.mjs';
import { getNormalizedEvent } from './normalize.ts';
import { makeTestTransport } from './transport.ts';

function getTestClient(
  callback: (event?: sentryTypes.Event) => void,
  integrations: sentryTypes.Integration[] = [],
): [Hub, DenoClient] {
  const client = new DenoClient({
    dsn: 'https://233a45e5efe34c47a3536797ce15dafa@nothing.here/5650507',
    debug: true,
    integrations: [...defaultIntegrations, ...integrations],
    stackParser: sentryUtils.createStackParser(sentryUtils.nodeStackLineParser()),
    transport: makeTestTransport(envelope => {
      callback(getNormalizedEvent(envelope));
    }),
  });

  const scope = new Scope();
  const hub = new Hub(client, scope);

  return [hub, client];
}

function delay(time: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, time);
  });
}

Deno.test('captureException', async t => {
  let ev: sentryTypes.Event | undefined;
  const [hub] = getTestClient(event => {
    ev = event;
  });

  function something() {
    return new Error('Some unhandled error');
  }

  hub.captureException(something());

  await delay(200);
  await assertSnapshot(t, ev);
});

Deno.test('captureMessage', async t => {
  let ev: sentryTypes.Event | undefined;
  const [hub] = getTestClient(event => {
    ev = event;
  });

  hub.captureMessage('Some error message');

  await delay(200);
  await assertSnapshot(t, ev);
});

Deno.test('App runs without errors', async _ => {
  const cmd = new Deno.Command('deno', {
    args: ['run', '--allow-net=some-domain.com', './test/example.ts'],
    stdout: 'piped',
    stderr: 'piped',
  });

  const output = await cmd.output();
  assertEquals(output.success, true);

  const td = new TextDecoder();
  const outString = td.decode(output.stdout);
  const errString = td.decode(output.stderr);
  assertEquals(outString, 'App has started\n');
  assertEquals(errString, '');
});
