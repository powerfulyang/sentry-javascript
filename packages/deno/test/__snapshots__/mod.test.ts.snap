export const snapshot = {};

snapshot[`captureException 1`] = `
{
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  exception: {
    values: [
      {
        mechanism: {
          handled: true,
          type: "generic",
        },
        stacktrace: {
          frames: [
            {
              colno: 20,
              filename: "ext:cli/40_testing.js",
              function: "outerWrapped",
              in_app: false,
              lineno: 472,
            },
            {
              colno: 33,
              filename: "ext:cli/40_testing.js",
              function: "exitSanitizer",
              in_app: false,
              lineno: 458,
            },
            {
              colno: 31,
              filename: "ext:cli/40_testing.js",
              function: "resourceSanitizer",
              in_app: false,
              lineno: 410,
            },
            {
              colno: 33,
              filename: "ext:cli/40_testing.js",
              function: "asyncOpSanitizer",
              in_app: false,
              lineno: 177,
            },
            {
              colno: 11,
              filename: "ext:cli/40_testing.js",
              function: "innerWrapped",
              in_app: false,
              lineno: 526,
            },
            {
              colno: 24,
              context_line: "  hub.captureException(something());",
              filename: "app:///test/mod.test.ts",
              function: "<anonymous>",
              in_app: true,
              lineno: 46,
              post_context: [
                "",
                "  await delay(200);",
                "  await assertSnapshot(t, ev);",
                "});",
                "",
                "Deno.test('captureMessage', async t => {",
                "  let ev: sentryTypes.Event | undefined;",
              ],
              pre_context: [
                "    ev = event;",
                "  });",
                "",
                "  function something() {",
                "    return new Error('Some unhandled error');",
                "  }",
                "",
              ],
            },
            {
              colno: 12,
              context_line: "    return new Error('Some unhandled error');",
              filename: "app:///test/mod.test.ts",
              function: "something",
              in_app: true,
              lineno: 43,
              post_context: [
                "  }",
                "",
                "  hub.captureException(something());",
                "",
                "  await delay(200);",
                "  await assertSnapshot(t, ev);",
                "});",
              ],
              pre_context: [
                "Deno.test('captureException', async t => {",
                "  let ev: sentryTypes.Event | undefined;",
                "  const [hub] = getTestClient(event => {",
                "    ev = event;",
                "  });",
                "",
                "  function something() {",
              ],
            },
          ],
        },
        type: "Error",
        value: "Some unhandled error",
      },
    ],
  },
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  timestamp: 0,
}
`;

snapshot[`captureMessage 1`] = `
{
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  level: "info",
  message: "Some error message",
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  timestamp: 0,
}
`;
